FLEXIBLE_INDEXES_NOTES.md

```diff
@@ -0,0 +1,124 @@
+
+Xarray should provide a minimal set of built-in index wrappers (this could be reduced to the indexes currently supported in Xarray, i.e., `pandas.Index` and `pandas.MultiIndex`). Other index wrappers may be implemented in 3rd-party libraries (recommended). The `XarrayIndex` base class should be part of Xarray's public API.
+
+#### 2.1.1 Index discoverability
+
+For better discoverability of Xarray-compatible indexes, Xarray could provide some mechanism to register new index wrappers, e.g., something like [xoak's `IndexRegistry`](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.IndexRegistry.html#xoak.IndexRegistry).
+
+### 2.2 Explicit vs. implicit index creation
+
+#### 2.2.1 Dataset/DataArray's `set_index` method
+
+New indexes can be built from an existing set of coordinates or variables in a Dataset/DataArray using the `.set_index()` method.
+
+TODO: describe API updates needed to provide the kind of index that we want to build from a set of coordinates.
+
+#### 2.2.2 Dataset/DataArray's `indexes` constructor argument
+
+TODO
+
+#### 2.2.3 Implicit default indexes
+
+In general explicit index creation should be preferred over implicit index creation. However, there is a majority of cases where basic `pandas.Index` objects could be built and used as indexes for 1-dimensional coordinates. For convenience, Xarray should automatically build such indexes for the coordinates where no index has been explicitly assigned in the Dataset/DataArray constructor or when indexes have been reset / dropped.
+
+For which coordinates?
+
+- A. only 1D coordinates with a name matching their dimension name
+- B. all 1D coordinates
+
+When to create it?
+
+- A. each time when a new Dataset/DataArray is created
+- B. only when we need it (i.e., when calling `.sel()` or `indexes`)
+
+#### 2.2.4 Implicit coordinates
+
+If we want to strictly follow the data model defined in [Section 1](#1_Data_Model), we need to ensure that at least one coordinate exists for each index (generally we need to ensure that the data indexed is properly exposed as one or more coordinates), especially when index objects are given directly to the Dataset/DataArray constructor ([Section 2.2.2](#222_DatasetDataArrays_indexes_constructor_argument)). This means that every `XarrayIndex` subclass must implement a method to get or generate one or more Xarray coordinates.
+
+Additionally, we might want to generate a coordinate that is specific to the index, generally with a composite data type, like a coordinate with n-elements tuple labels for a n-levels `pandas.MultiIndex`. This should be optional, though. An `XarrayIndex` may or may not support this. This coordinate probably shouldn't be created implicitly or by default when setting the new index. Also, this might raise some issues, e.g., what would be the name of this coordinate? For a `pandas.MultiIndex` it would make sense in many (but not all) cases to choose the dimension name as the coordinate name.
```

Comment:
This is what we do right now, but I'm not sure it's really a good idea.

In the MultiIndex case, I would lean towards a single XarrayIndex object holding the MultiIndex, and a number of 1D coordinate variables for the levels of the MultiIndex.

Reply:
I agree, I'm not sure either it's a good idea. I haven't had any use case for this, but maybe other users have some? We could still manually create new coordinates from an index object, though.

I'm wondering if this would be temporarily useful for backward compatibility. For example, if we want to still support `da.set_index(dim_x=['coord_a', 'coord_b'])`, would it be ok not having any `dim_x` coordinate in the resulting DataArray?